### PR TITLE
PartialMergeJoin flush data on disk

### PR DESCRIPTION
**Pull request by 4ertus2**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category (leave one):
- Improvement

Short description (up to few sentences):
Flush parts of right-hand joining table on disk in PartialMergeJoin (if there is not enough memory). Load data back when needed.

### CODE REVIEW

File: dbms/src/Interpreters/MergeJoin.cpp [hunk 1]
```diff
@@ -546,4 +725,30 @@ void MergeJoin::addRightColumns(Block & block, MutableColumns && right_columns)
...
+template <bool in_memory>
+size_t MergeJoin::rightBlocksCount()
+{
+    if constexpr (!in_memory)
+        return flushed_right_blocks.size();
+    else
+        return loaded_right_blocks.size();
+}
+
+template <bool in_memory>
```
Comment (akuzm) [inline dbms/src/Interpreters/MergeJoin.cpp]: Why is it a template parameter and not just a normal parameter? At a first glance, it looks like this function works on the level of entire blocks, so we can just branch in runtime and don't have to statically generate two functions, and it won't lead to any performance problems.

File: dbms/src/Interpreters/MergeJoin.cpp [hunk 2]
```diff
@@ -320,21 +326,127 @@ void joinInequalsLeft(const Block & left_block, MutableColumns & left_columns, M
...
+        flushStreamToFiles(path, sample_block, sorted_input, sorted_blocks);
+    }
+
+    sorted_files.emplace_back(std::move(sorted_blocks));
+    if (sorted_files.size() >= max_size)
+        merge();
+}
+
+/// TODO: better merge strategy
+void MiniLSM::merge(std::function<void(const Block &)> callback)
```
Comment (akuzm) [inline dbms/src/Interpreters/MergeJoin.cpp]: Can we have a more descriptive name for the callback, like `on_before_flush_block`?

**Comment by akuzm** [comment]
Overall looks good to me.

File: dbms/src/Interpreters/MergeJoin.cpp (hunk 1 above)
Comment (4ertus2) [reply dbms/src/Interpreters/MergeJoin.cpp]: It's some technics to use template for passthrought strategy without additional cost. If I remember correctly it's "Template Method" without interfaces.
```
tempalte <int strategy>
dificult_func(...)
{
    x = simple_func1<strategy>()
    ....
    simple_func2<strategy>()
}
```

Comment (akuzm) [reply dbms/src/Interpreters/MergeJoin.cpp]: Yep, that's how I read it. But it's not without cost -- it increases code size and compile times. Personally, I also find it more cumbersome to read. This code doesn't have strong performance requirements, and the strategy is a simple bool, so maybe it doesn't have to pass it statically, and passing dinamically will suffice?

Comment (4ertus2) [reply dbms/src/Interpreters/MergeJoin.cpp]: > This code doesn't have strong performance requirements
Not exactly. On-disk variant really is not perf critical here. The idea is to have zero cost for in-memory variant. It should be as fast as there's not on-disk variant at all. It has very simple loadRightBlock() that expected to be slower with additional if.

Comment (akuzm) [reply dbms/src/Interpreters/MergeJoin.cpp]: Well, OK then.

**Comment by akuzm** [comment]
The performance test not finishing is related to problems in master. This PR shouldn't have performance consequences, since it's for a new type of join. Will merge as is.

**Pull request merged**